src/cargo/sources/git/oxide.rs

```diff
@@ -0,0 +1,338 @@
+    })
+}
+
+fn translate_progress_to_bar(
+    progress_bar: &mut Progress<'_>,
+    root: Weak<gix::progress::tree::Root>,
+) -> CargoResult<()> {
+    let read_pack_bytes: gix::progress::Id =
+        gix::odb::pack::bundle::write::ProgressId::ReadPackBytes.into();
+    let delta_index_objects: gix::progress::Id =
+        gix::odb::pack::index::write::ProgressId::IndexObjects.into();
+    let resolve_objects: gix::progress::Id =
+        gix::odb::pack::index::write::ProgressId::ResolveObjects.into();
+
+    // We choose `N=10` here to make a `300ms * 10slots ~= 3000ms`
+    // sliding window for tracking the data transfer rate (in bytes/s).
+    let mut last_update = Instant::now();
+    let mut counter = MetricsCounter::<10>::new(0, last_update);
+
+    let mut tasks = Vec::with_capacity(10);
+    let update_interval = std::time::Duration::from_millis(300);
+    let short_check_interval = Duration::from_millis(50);
+
+    while let Some(root) = root.upgrade() {
+        let not_yet = last_update.elapsed() < update_interval;
+        if not_yet {
+            std::thread::sleep(short_check_interval);
+        }
+        root.sorted_snapshot(&mut tasks);
+
+        fn progress_by_id(
+            id: gix::progress::Id,
+            task: &gix::progress::Task,
+        ) -> Option<&gix::progress::Value> {
+            (task.id == id).then(|| task.progress.as_ref()).flatten()
+        }
+        fn find_in<K>(
+            tasks: &[(K, gix::progress::Task)],
+            cb: impl Fn(&gix::progress::Task) -> Option<&gix::progress::Value>,
+        ) -> Option<&gix::progress::Value> {
+            tasks.iter().find_map(|(_, t)| cb(t))
+        }
+
+        if let Some(objs) = find_in(&tasks, |t| progress_by_id(resolve_objects, t)) {
+            // Resolving deltas.
+            let objects = objs.step.load(Ordering::Relaxed);
+            let total_objects = objs.done_at.expect("known amount of objects");
+            let msg = format!(", ({objects}/{total_objects}) resolving deltas");
+
+            progress_bar.tick(objects, total_objects, &msg)?;
+        } else if let Some((objs, read_pack)) =
+            find_in(&tasks, |t| progress_by_id(read_pack_bytes, t)).and_then(|read| {
+                find_in(&tasks, |t| progress_by_id(delta_index_objects, t))
+                    .map(|delta| (delta, read))
+            })
+        {
+            // Receiving objects.
+            let objects = objs.step.load(Ordering::Relaxed);
+            let total_objects = objs.done_at.expect("known amount of objects");
+            let received_bytes = read_pack.step.load(Ordering::Relaxed);
+
+            let now = Instant::now();
+            if !not_yet {
+                counter.add(received_bytes, now);
+                last_update = now;
+            }
+            let (rate, unit) = human_readable_bytes(counter.rate() as u64);
+            let msg = format!(", {rate:.2}{unit}/s");
+
+            progress_bar.tick(objects, total_objects, &msg)?;
```

Comment:
(Not really a blocker. We can fix this in follow-ups I believe.)

There are still some subtle changes/regressions regarding the progress bar.

* In the old implementation, we didn't split download and resolving deltas into two steps; they shared the one progress. In this PR we have different upper bounds for each task. Is it possible to follow the old behavior?
* It seems like the `delta_index_objects` task always gets stuck in 50%. Is that a bug in gix or just we display the wrong thing?

I don't mind if we eventually get two progresses but the progress reports should be correct and make sense to users.

Here is a record you can take as a reference: <https://asciinema.org/a/563786?t=3>

Reply:
A great catch, I just (tried) to go by keeping the code similar and called it a day, but didn't actually compare if it is the same.

> In the old implementation, we didn't split download and resolving deltas into two steps; they shared the one progress. In this PR we have different upper bounds for each task. Is it possible to follow the old behavior?

That should be possible, I will give it a another shot.

> It seems like the delta_index_objects task always gets stuck in 50%. Is that a bug in gix or just we display the wrong thing?

I think it's a display issue, as it doesn't get stuck when cloning with `gix` as far as I can tell. Maybe it's fixed as soon as I take a closer look at the above issue.

> Here is a record you can take as a reference: https://asciinema.org/a/563786?t=3

Whoop whoop, that will be so helpful, thanks a lot!

I will be back once this is fixed, with a video for comparison :).